fix: Apply offsets in DOM that exists before the polyfill is loaded

smockle opened the pull request:
This PR:

1. Removes `type="module"` from the “Usage” instructions in the README, since the polyfill script isn’t a module.

2. Applies offsets immediately to existing DOM (this complements the existing `MutationObserver`s) and adds a test to validate this behavior.

3. Adds tests to check various ways the polyfill may be loaded on a page.

I’m still thinking about how to handle shadow roots; that’ll be a separate PR.

Kudos @bkardell for raising this issue.

bkardell commented on README.md:
````diff
@@ -9,7 +9,7 @@ Polyfill for the `headingoffset` attribute. Related discussion in https://github
 
 ```html
-<script type="module" src="headingoffset-polyfill.js"></script>
+<script src="headingoffset-polyfill.js"></script>
````
I'm not sure what would your usage recommendation be? Do you have one?  If you include it in the head in blocking fashion like this it will have to fully download and execute and therefore the handlers will be there - but they'll be pretty busy if it is a big page downloading then.  You could also do it like this at the end but then you have to do a full document scan and if it is a big document there could be some time where the headings are wrong.  Maybe an async attribute is a nice compromise? Maybe just add a note to the effect that the script can be used in blocking or non-blocking fashion?

smockle replied:
> …what would your usage recommendation be? Do you have one?

No, I don’t have a strong opinion on where folks should put the script tag (and/or whether they should use `defer`, `async`, etc.)

> Maybe just add a note to the effect that the script can be used in blocking or non-blocking fashion?

Sure, I’m open to adding a note outlining options and trade-offs. Is there a note like this in another polyfill’s README that I could read over for wording/formatting inspiration?

bkardell replied:
No, I don't have any examples I can think of - but I always wonder about this exact thing when I find a polyfill. Maybe just "The script will scan the existing document and monitor for changes after that" and people can figure the rest out for themselves :). I wouldn't block anything based on this comment btw... I would use async probably on my site :)

smockle replied:
Thanks! Sounds good to me; I added that note via https://github.com/smockle/headingoffset-polyfill/pull/16/commits/32887ec19affac8886427ee86910c6cdd3055924.

bkardell commented:
Wait it isn't captured in here but you will throw on `.observe(document.body,...` in the event that you include this in the head...

smockle commented:
> Wait it isn't captured in here but you will throw on `.observe(document.body,...` in the event that you include this in the head...

Nice catch! I added tests for this case in https://github.com/smockle/headingoffset-polyfill/pull/16/commits/3aeb3f75ee36ac437b2a93078091bce002d26906, then I fixed the issue (by observing `document.documentElement` instead) in https://github.com/smockle/headingoffset-polyfill/pull/16/commits/64d7fd0051cfa8a0a5b40c1972c321b52a5db569. The expanded test suite passes.

bkardell commented:
Ah yes... or that :) That is probably easier, it means you'll process a few more nodes maybe, but 🤷 probably fine.
